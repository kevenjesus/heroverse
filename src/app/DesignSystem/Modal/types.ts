import { ReactNode } from "react";

export interface ModalProps {
    children: ReactNode
    title?: string
    open: boolean
    onClose: () => void
    onClosed?: () => void
    onOpened?: () => void
}