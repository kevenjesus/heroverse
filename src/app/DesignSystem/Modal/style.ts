import { breakpoints } from '@/app/styles/guide'
import styled from 'styled-components'


export const ModalItem = styled.div`
    width: 95%;
    padding: 20px;
    background: #fff;
    position: fixed;
    top: 10%;
    left: 2.5%;
    z-index: 6;
    @media (min-width: ${breakpoints.sm}) {
        width: 450px;
        left: 50%;
        margin-left: -225px;
        top: 20%;
    }
`

export const Heading = styled.header`
    width: 100%;
    height: 40px;
    margin-bottom: 20px;
    display: flex;
    justify-content: space-between;
    align-items: center;
    button {
        margin-left: auto;
    }
`