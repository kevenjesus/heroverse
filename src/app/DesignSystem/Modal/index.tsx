import { useCallback, useEffect, useState } from 'react'
import Headline from '../Headline'
import * as S from './style'
import { ModalProps } from './types'

export default function Modal({children, title, open, onClose, onClosed, onOpened}: ModalProps) {
    const [isClosed, setClosed] = useState(false)

    const onClosedModal = useCallback(() => {
        onClose()
        setClosed(true)
    }, [])

    useEffect(() => {
        if(open) {
            setClosed(false)
            onOpened && onOpened()
        }
    }, [open])
    
    useEffect(() => {
        if(isClosed) {
            onClosed && onClosed()
        }
    }, [isClosed])

    if(!open) {
        return <></>
    }
    return (
        <>
            <S.Overlay />
            <S.ModalItem>
                <S.Heading>
                    <Headline as="h3">{title}</Headline>
                    <button onClick={onClosedModal} role='button' aria-label='close-modal'>
                        <img src='/cancel.png' alt="fechar" />
                    </button>
                </S.Heading>
                {children}
            </S.ModalItem>
        </>
    )
}