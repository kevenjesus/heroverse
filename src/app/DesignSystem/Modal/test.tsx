import { cleanup, fireEvent, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom'
import Modal from '.';

let stateModal = true
const onClosed = jest.fn()
const onOpened = jest.fn()
const onCloseModal = jest.fn(() => stateModal = false)


describe('Modal', () => {
  afterEach(() => {
    cleanup()
  }) 
  it('should render', () => {
    const { container } = render(<Modal open={stateModal} onClose={onCloseModal} title="modal">modal</Modal>);
    expect(container).toMatchSnapshot()
  });

  it('should not render', () => {
    const { container } = render(<Modal open={false} onClose={onCloseModal} title="modal">modal</Modal>);
    expect(container).toMatchSnapshot()
  });

  it('should on event close modal', async () => {
    render(<Modal open={stateModal} onClose={onCloseModal} onClosed={onClosed} title="modal">modal</Modal>);
    const buttonClose = await screen.findByRole('button', { name: 'close-modal'})
    fireEvent.click(buttonClose)
    expect(onClosed).toBeCalled()
  });

  it('should on event opened modal', async () => {
    render(<Modal open={true} onClose={onCloseModal} onOpened={onOpened} title="modal">modal</Modal>);
    await waitFor(() => expect(onOpened).toBeCalled())
  });
});
