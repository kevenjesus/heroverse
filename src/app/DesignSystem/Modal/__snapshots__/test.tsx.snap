// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should not render 1`] = `<div />`;

exports[`Modal should render 1`] = `
<div>
  <div
    class="sc-beqWaB hPQajf"
  />
  <div
    class="sc-hLseeU dxqZKz"
  >
    <header
      class="sc-eDDNvR euiAjC"
    >
      <h3
        class="sc-dmqHEX fohnnB"
        type="h3"
      >
        modal
      </h3>
      <button
        aria-label="close-modal"
        role="button"
      >
        <img
          alt="fechar"
          src="/cancel.png"
        />
      </button>
    </header>
    modal
  </div>
</div>
`;
